---
title: "BotanicalCorrection"
author: "Vincyane Badouard"
date: "11/04/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Trouver des packages de : détéction de special characters, pour réccupérer la famille à partir du scientific name, et de update selon phylogenetic classification

Return: ... and fill the 'Comment' column

Details:
- No special characters (typography) (*TNRS package*)  
- Check invariant botanical informations per IdTree (1 IdTree = 1 family, 1 scientific and 1 vernacular name)  
- No family name in the genus and species columns (detection of the suffix "aceae" in the genus and species columns (it is specific to the family name) -> Comment (+transfer to the Family column?) (sort(unique(data[grep("aceae", data$genus), genus]))))  
- Family & Scientific names match (*BIOMASS::getTaxonomy*)
- Scientific & vernacular names match
- Update the scientific botanical names with the current phylogenetic classification

+ **add the correction source!**

Tests:
- No special characters
- Invariant botanical informations per IdTree
- No family name in the genus and species columns 
- Family & Scientific names match
- Scientific & vernacular names match
- Update the scientific botanical names with the current phylogenetic classification


# Packages libraries
```{r, include = FALSE}
library(TreeData)
library(data.table)
library(ggplot2)
library(Taxonstand)
```

# Import data & other inputs
```{r}
data("TestData")
setDT(TestData) # data.frame to data.table

# Args

# ScientificVernaTable ou BotaMatchTable # Scientific-vernacular names match table (data.frame)
```

# Create data with different type of botanical error
- No special characters
- Invariant botanical informations per IdTree
- No family name in the genus and species columns 
- Family & Scientific names match
- Scientific & vernacular names match
- Old scientific name
```{r}
# Fabaceae Dicorynia guianensis (angelique) "a"
# Lecythidaceae Eschweilera sagotiana (maho noir) "b"
# Chrysobalanaceae Licania alba (koko) "c"
# Fabaceae Eperua falcata (wapa) "d"
Data <- data.table(Site = "Nowhere",
                   IdTree = c(rep("a", 4), rep("b", 4), rep("c", 4), rep("d", 4)), # 4 ind
                   Year = rep(c(2000:2003), 4) # 4 years each
)
Data <- Data[order(IdTree, Year)]
Data[, Family := c(rep("Fabaceae", 4), rep("Lecythidaceae", 4), rep("Chrysobalanaceae", 4), rep("Fabaceae", 4))]
Data[, Genus := c(rep("Dicorynia", 4), rep("Eschweilera", 4), rep("Licania", 4), rep("Eperua", 4))]
Data[, Species := c(rep("guianensis", 4), rep("sagotiana", 4), rep("alba", 4), rep("falcata", 4))]
Data[, VernName := c(rep("angelique", 4), rep("maho noir", 4), rep("koko", 4), rep("wapa", 4))]

# Create errors
## Special characters
Data[IdTree %in% "a" & Year %in% 2003, ("VernName") := "Angélique"] # good answer: "angelique"
Data[IdTree %in% "a", ("Family") := "Fabacé"] # good answer: "Fabaceae"
Data[IdTree %in% "c" & Year %in% 2003, ("Genus") := "Licanï_a"] # good answer: "Licania"

## Variant botanical informations per IdTree
Data[IdTree %in% "d" & Year %in% 2002, ("Species") := "grandiflora"] # good answer: "falcata"

## Family name in the genus/species columns
Data[IdTree %in% "b", ("Species") := "Lecythidaceae"] # good answer: "sagotiana"

## Family & Scientific names unmatch
Data[IdTree %in% "c", ("Family") := "Lecythidaceae"] # good answer: "Chrysobalanaceae"

## Scientific & vernacular names unmatch
Data[IdTree %in% "d" & Year %in% 2003, ("VernName") := "leaf"]  # good answer: "wapa"

## Old scientific name (A trouver)

Data[, ScientificName := paste(Genus, Species)]

```

# Ask to the user

# Check args

# Special characters (Typographie)
TNRS package é, è or œ
Reurn: A data.frame with the following components:

$Taxon $Genus $Species $Authority
Original taxon/genus/Species/author name as provided in input.

$Hybrid.marker (*to delete*)
Hybrid marker, if taxon is indicated as hybrid in the input.

$Abbrev	(*to delete*)
Original abbreviation other than infraspecific rank included in input taxon, including "cf.", "aff.", "agg.", "nom. cons.", "nom. cons. prop.", "nom. inval.", "s.l.", and "s.str." and their orthographic variants.

$Infraspecific.rank	(*to delete*)
Original infraspecific rank abbreviation as provided in input, including "subsp.", "var.", "f.", and their orthographic variants.

$Infraspecific	(*to delete*)
Original infraspecific epithet as provided in input. If infra = FALSE, this is not shown.

$ID	
The Plant List record ID of the matched taxon before resolving synonyms.

$Plant.Name.Index	
Logical.*If TRUE the name is in TPL*. If a taxon at infraspecific level is not in TPL, Plant.Name.Index equals FALSE, except for nominal infraspecies. Also compare Higher.level.

$TPL.version	
Version of TPL used.

$Taxonomic.status	
Taxonomic status of the matched taxon in TPL, either 'Accepted', 'Synonym', 'Unresolved', or 'Misapplied'.

$Family	$New.Genus $New.Species	$New.Infraspecific	$New.Authority	
Family/Genus/Species/Infraspecific/Author name, extracted from TPL for the valid form of the taxon.


$New.Hybrid.marker (*to delete*)
Hybrid marker, extracted from TPL for the valid form of the taxon.

$New.Infraspecific.rank	
Infraspecific rank abbreviation, extracted from TPL for the valid form of the taxon, including "subsp.", "var." and "f.".

$New.ID	
The Plant List record ID of the taxon, once synonyms have been replaced by valid names. For accepted and unresolved names, this field will be equivalent to ID.

$New.Taxonomic.status	
Taxonomic status of the resolved taxon in TPL, once synonyms have been replaced by valid names. 'Accepted' or 'Unresolved'.

$Typo	
Logical. *If TRUE there was a spelling error in the specific or infraspecific epithet* that has been corrected.

$WFormat	
Logical. If TRUE, fields in TPL had the wrong format for information to be automatically extracted as they were not properly tabulated or, alternatively, there was not a unique solution.

$Higher.level	
Logical. If TRUE, the input taxon is at infraspecific level and does not occur in TPL, and the higher (species) level is provided in the output instead. Also see Plant.Name.Index.

$Date	(*to delete*)
Current date according to Sys.Date
```{r}
# The Plant List
RsltSfc <- Taxonstand::TPL(Data$ScientificName, diffchar = 2, ) # diffchar: maximum difference of characters nbr between input and output

RsltGS <- Taxonstand::TPL(genus = Data$Genus, species = Data$Species, infra = FALSE, diffchar = 2, ) # diffchar: maximum difference of characters nbr between input and output

setDT(RsltSfc) # df to dt
setDT(RsltGS) # df to dt

RsltSfc[, c("Hybrid.marker", "New.Hybrid.marker", # No hybrid
            "Abbrev", "Infraspecific.rank", "Infraspecific", "New.Infraspecific.rank", "New.Infraspecific", "Higher.level", # No infrasp
            "Date") := NULL]

# Autre fonction : https://github.com/EnquistLab/RTNRS/blob/master/R/TNRS_base.R
# un shiny : https://github.com/APonce73/ShinyR-Taxonomic-correction
```



# Check invariant botanical informations per IdTree
"Family", "ScientificName", VernName"
```{r}
duplicated_ID <- CorresIDs <- vector("character")

# For each site
for (s in unique(na.omit(Data$Site))) {
  
  BotaIDCombination <- na.omit(unique(
    Data[Data$Site == s, .(IdTree, Family, ScientificName, VernName)]
  ))
  
  CorresIDs <- BotaIDCombination[, IdTree] # .(IdTree) all the Idtree's having a unique X-Yutm) combination
  
  if(!identical(CorresIDs, unique(CorresIDs))){ # check if it's the same length, same ids -> 1 asso/ID
    
    duplicated_ID <- unique(CorresIDs[duplicated(CorresIDs)]) # identify the Idtree(s) having several P-SubP-TreeFieldNum combinations
    
    Data <- GenerateComment(Data,
                            condition =
                              Data[,Site] == s
                            & Data[,IdTree] %in% duplicated_ID,
                            comment = "Different botanical informations (Family, ScientificName or VernName) for a same IdTree")
  }
} # end site loop

unique(Data[IdTree %in% duplicated_ID,
            .(IdTree = sort(IdTree), Family, ScientificName, VernName, Comment)]) # to check 

```

# No family name in the genus and species columns
(detection of the suffix "aceae" in the genus and species columns (it is specific to the family name) -> Comment (+transfer to the Family column?)
```{r}
sort(unique(Data[grep("aceae", Data$ScientificName), ScientificName]))
sort(unique(Data[grep("aceae", Data$Genus), Genus]))
sort(unique(Data[grep("aceae", Data$Species), Species]))


```


# Family & Scientific names match
(*BIOMASS::getTaxonomy*)

# Scientific & vernacular names match

# Update the scientific botanical names with the current phylogenetic classification

