---
title: "BotanicalCorrection"
author: "Vincyane Badouard"
date: "11/04/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Trouver des packages de : 
+ détéction de **special characters**,
+ pour **réccupérer la famille** à partir du scientific name,
+ et de **update** selon phylogenetic classification

Return: ... and fill the 'Comment' column

Details:
- No special characters (typography) (*TNRS package*)
- No family name in the genus and species columns (detection of the suffix "aceae" in the genus and species columns (it is specific to the family name) -> Comment (+transfer to the Family column?) (sort(unique(data[grep("aceae", data$genus), genus]))))  
- Correct spelling of botanical names (*Taxonstand::TPL*)
- Check **invariant botanical informations per IdTree** (1 IdTree = 1 family, 1 scientific and 1 vernacular name)  
- Family & Scientific names match (*BIOMASS::getTaxonomy*)
- Scientific & vernacular names match
- Update the scientific botanical names with the current phylogenetic classification

+ **add the correction source!** : 'BotanicalCorrectionSource' column

Tests:
- No special characters
- Invariant botanical informations per IdTree
- No family name in the genus and species columns 
- Family & Scientific names match
- Scientific & vernacular names match
- Update the scientific botanical names with the current phylogenetic classification


# Packages libraries
```{r, include = FALSE}
library(TreeData)
library(data.table)
library(ggplot2)
library(Taxonstand)
```

# Import data & other inputs
```{r}
data("TestData")
setDT(TestData) # data.frame to data.table

# Args

# ScientificVernaTable ou BotaMatchTable # Scientific-vernacular names match table (data.frame)
```

# Create data with different type of botanical error
- No special characters (!"#$%&’()*+,-./:;<=>?@[]^_`{|}~)
- No orthographical error
- Invariant botanical informations per IdTree
- No family name in the genus and species columns 
- Family & Scientific names match
- Scientific & vernacular names match
- Old scientific name
```{r}
# Fabaceae Dicorynia guianensis (angelique) "a"
# Lecythidaceae Eschweilera sagotiana (maho noir) "b"
# Chrysobalanaceae Licania alba (koko) "c"
# Fabaceae Eperua falcata (wapa) "d"
Data <- data.table(Site = "Nowhere",
                   IdTree = c(rep("a", 4), rep("b", 4), rep("c", 4), rep("d", 4)), # 4 ind
                   Year = rep(c(2000:2003), 4) # 4 years each
)
Data <- Data[order(IdTree, Year)]
Data[, Family := c(rep("Fabaceae", 4), rep("Lecythidaceae", 4), rep("Chrysobalanaceae", 4), rep("Fabaceae", 4))]
Data[, Genus := c(rep("Dicorynia", 4), rep("Eschweilera", 4), rep("Licania", 4), rep("Eperua", 4))]
Data[, Species := c(rep("guianensis", 4), rep("sagotiana", 4), rep("alba", 4), rep("falcata", 4))]
Data[, VernName := c(rep("angelique", 4), rep("maho noir", 4), rep("koko", 4), rep("wapa", 4))]

# Create errors

## Missing value in  Family, ScientificName/Genus, species, VernName
Data[IdTree %in% "d" & Year %in% 2000, ("Family") := NA_character_]
Data[IdTree %in% "b" & Year %in% 2000, ("Genus") := NA_character_]
Data[IdTree %in% "b", ("Family") := NA_character_]
Data[IdTree %in% "c" & Year %in% 2000, ("Species") := NA_character_]
Data[IdTree %in% "d" & Year %in% 2000, ("VernName") := NA_character_]



## Special characters
Data[IdTree %in% "a" & Year %in% 2003, ("VernName") := "Angélique"] # good answer: "angelique"
Data[IdTree %in% "a", ("Family") := "Fabacé"] # good answer: "Fabaceae"
Data[IdTree %in% "c" & Year %in% 2003, ("Genus") := "Licanï_a"] # good answer: "Licania"

## Variant botanical informations per IdTree
Data[IdTree %in% "d" & Year %in% 2002, ("Species") := "grandi!flora"] # good answer: "falcata"

## Family name in the genus/species columns
Data[IdTree %in% "b", ("Species") := "Lecythidaceae"] # good answer: "sagotiana"

## Family & Scientific names unmatch
Data[IdTree %in% "c", ("Family") := "Lecythidaceae"] # good answer: "Chrysobalanaceae"

## Scientific & vernacular names unmatch
Data[IdTree %in% "d" & Year %in% 2003, ("VernName") := "leaf"]  # good answer: "wapa"

## Old scientific name (A trouver)

Data[, ScientificName := paste(Genus, Species)]

```

# Ask to the user

# Check args

# Corrected columns initialisation
```{r}
Data[, FamilyCor := Family]
Data[, GenusCor := Genus]
Data[, SpeciesCor := Species]
```


# Missing value
Family, ScientificName/Genus, species, VernName
```{r}
Vars <- c("Family", "ScientificName", "Genus", "Species", "VernName")

for (v in 1:length(Vars)) {
  
  if(Vars[v] %in% names(Data)){ # If the column exists
    
    Data <- GenerateComment(Data,
                            condition = is.na(Data[,get(Vars[v])]),
                            comment = paste0("Missing value in ", Vars[v]))
  }
}

# Data[Comment != ""] # to check 
```


# Special characters
Remove : !"#$%&’()*+,-./:;<=>?@[]^_`{|}~
```{r}

Data[, FamilyCor := gsub("[[:punct:]]", "", Data$FamilyCor)]
Data[, GenusCor := gsub("[[:punct:]]", "", Data$GenusCor)]
Data[, SpeciesCor := gsub("[[:punct:]]", "", Data$SpeciesCor)]

# Comment :

Data <- GenerateComment(Data,
                        condition = grepl('[[:punct:]]', Data$Family), # TRUE if there are any special character
                        comment = "Special characters in the 'Family'")

Data <- GenerateComment(Data,
                        condition = grepl('[[:punct:]]', Data$Genus), # TRUE if there are any special character
                        comment = "Special characters in the 'Genus'")

Data <- GenerateComment(Data,
                        condition = grepl('[[:punct:]]', Data$Species), # TRUE if there are any special character
                        comment = "Special characters in the 'Species'")

```

# No family name in the genus and species columns
(detection of the suffix "aceae" in the genus and species columns (it is specific to the family name) -> Comment (+transfer to the Family column?)
```{r}

Data[grep("aceae", Data$Genus), `:=`(FamilyCor = ifelse(is.na(FamilyCor), Genus, FamilyCor), GenusCor = NA_character_)]
Data[grep("aceae", Data$Species), `:=`(FamilyCor = ifelse(is.na(FamilyCor), Species, FamilyCor), SpeciesCor = NA_character_)]
# sort(unique(Data[grep("aceae", Data$ScientificName), ScientificName]))

# Comment : 
Data <- GenerateComment(Data,
                        condition = grepl("aceae", Data$Genus) | grepl("aceae", Data$Species),
                        comment = "Names ending in 'aceae' cannot be genus or species names")


```


# Orthographical error
TNRS package é, è or œ
Reurn: A data.frame with the following components:

$Taxon $Genus $Species $Authority
Original taxon/genus/Species/author name as provided in input.

$Hybrid.marker (*to delete*)
Hybrid marker, if taxon is indicated as hybrid in the input.

$Abbrev	(*to delete*)
Original abbreviation other than infraspecific rank included in input taxon, including "cf.", "aff.", "agg.", "nom. cons.", "nom. cons. prop.", "nom. inval.", "s.l.", and "s.str." and their orthographic variants.

$Infraspecific.rank	(*to delete*)
Original infraspecific rank abbreviation as provided in input, including "subsp.", "var.", "f.", and their orthographic variants.

$Infraspecific	(*to delete*)
Original infraspecific epithet as provided in input. If infra = FALSE, this is not shown.

$ID	
The Plant List record ID of the matched taxon before resolving synonyms.

$Plant.Name.Index	
Logical.*If TRUE the name is in TPL*. If a taxon at infraspecific level is not in TPL, Plant.Name.Index equals FALSE, except for nominal infraspecies. Also compare Higher.level.

$TPL.version	
Version of TPL used.

$Taxonomic.status	
Taxonomic status of the matched taxon in TPL, either 'Accepted', 'Synonym', 'Unresolved', or 'Misapplied'.

$Family	$New.Genus $New.Species	$New.Infraspecific	$New.Authority	
Family/Genus/Species/Infraspecific/Author name, extracted from TPL for the valid form of the taxon.


$New.Hybrid.marker (*to delete*)
Hybrid marker, extracted from TPL for the valid form of the taxon.

$New.Infraspecific.rank	
Infraspecific rank abbreviation, extracted from TPL for the valid form of the taxon, including "subsp.", "var." and "f.".

$New.ID	
The Plant List record ID of the taxon, once synonyms have been replaced by valid names. For accepted and unresolved names, this field will be equivalent to ID.

$New.Taxonomic.status	
Taxonomic status of the resolved taxon in TPL, once synonyms have been replaced by valid names. 'Accepted' or 'Unresolved'.

$Typo	
Logical. *If TRUE there was a spelling error in the specific or infraspecific epithet* that has been corrected.

$WFormat	
Logical. If TRUE, fields in TPL had the wrong format for information to be automatically extracted as they were not properly tabulated or, alternatively, there was not a unique solution.

$Higher.level	
Logical. If TRUE, the input taxon is at infraspecific level and does not occur in TPL, and the higher (species) level is provided in the output instead. Also see Plant.Name.Index.

$Date	(*to delete*)
Current date according to Sys.Date

Autre fonction : https://github.com/EnquistLab/RTNRS/blob/master/R/TNRS_base.R
un shiny : https://github.com/APonce73/ShinyR-Taxonomic-correction (marche pas)

```{r}
Data[, ScientificNameCor := paste(GenusCor, SpeciesCor)]

## with Scientific name (25 var)
TPLCor <- Taxonstand::TPL(unique(Data$ScientificNameCor), corr = TRUE, diffchar = 20, max.distance = 1) # diffchar: maximum difference of characters nbr between input and output
# with Genus and species marche pas bien pcq décale genre et sp quand on unique())

setDT(TPLCor) # df to dt

# Take only corrected names. Columns: New.Genus, New.Species, Family, Typo
TPLCor <- TPLCor[New.Genus != Genus | New.Species != Species,]
TPLCor <- TPLCor[, list(Taxonomic.status, Typo, Family, Taxon, New.Genus, New.Species)]
TPLCor <- setnames(TPLCor, "Family", "New.Family") # rename columns
TPLCor[, BotanicalCorrectionSource := "The Plant List"] # create the Source


# Join the corrected Genus and Species, by original 'ScientificNameCor'
Data <- merge(Data, TPLCor, by.x = "ScientificNameCor", by.y = "Taxon", all.x = TRUE) #  by.x = "ScientificNameCor", by.y = "Taxon", sort = FALSE

# Update correction columns
Data[, GenusCor := ifelse(!is.na(New.Genus), New.Genus, GenusCor)]
Data[, SpeciesCor := ifelse(!is.na(New.Species), New.Species, SpeciesCor)]


# Comment: AREETE LA
## if "Synonym" : 
Data <- GenerateComment(Data,
                        condition = Data$Taxonomic.status == "Synonym",
                        comment = "'ScientificName' is a synonym of the accepted botanical name")
## if Typo == TRUE : 
Data <- GenerateComment(Data,
                        condition = Data$Typo == TRUE,
                        comment = "Spelling error in the 'ScientificName'")

# Remove columns that have become useless
Data[, c("Taxonomic.status", "Typo", "New.Genus", "New.Species") := NULL]

# Expand Family info (A FAIRE, je sais pas encore comment)
```



# Check invariant botanical informations per IdTree ARETE LA !
"Family", "ScientificName", VernName"
```{r}
duplicated_ID <- CorresIDs <- vector("character")

# For each site
for (s in unique(na.omit(Data$Site))) {
  
  BotaIDCombination <- na.omit(unique(
    Data[Data$Site == s, .(IdTree, Family, ScientificName, VernName)]
  ))
  
  CorresIDs <- BotaIDCombination[, IdTree] # .(IdTree) all the Idtree's having a unique X-Yutm) combination
  
  if(!identical(CorresIDs, unique(CorresIDs))){ # check if it's the same length, same ids -> 1 asso/ID
    
    duplicated_ID <- unique(CorresIDs[duplicated(CorresIDs)]) # identify the Idtree(s) having several P-SubP-TreeFieldNum combinations
    
    Data <- GenerateComment(Data,
                            condition =
                              Data[,Site] == s
                            & Data[,IdTree] %in% duplicated_ID,
                            comment = "Different botanical informations (Family, ScientificName or VernName) for a same IdTree")
  }
} # end site loop

unique(Data[IdTree %in% duplicated_ID,
            .(IdTree = sort(IdTree), Family, ScientificName, VernName, Comment)]) # to check 

```


# Family & Scientific names match
Recovering the Family name by Genus
(*BIOMASS::getTaxonomy*) with APG III family
```{r}
FamilyData <- 
  setDT( # as data.table
    BIOMASS::getTaxonomy(unique(Data$GenusCor), findOrder = FALSE)
  )

FamilyData <- setnames(FamilyData, "family", "FamilyCor") # rename columns


Data <- merge(Data, FamilyData, by.x = "GenusCor", by.y = "inputGenus",  all.x = TRUE, sort = FALSE)

# Generate a comment if the family name is incorrect
Data <- GenerateComment(Data,
                        condition = Data[,Family] != Data[,FamilyCor],
                        comment = "The 'Family' name is incorrect")

Data[Family != FamilyCor, FamilyCorSource := "APG III family"] # create the Source

# Comment je gère ces différentes colonnes ? ARRETEE ICI
```


# Scientific & vernacular names match

# Update the scientific botanical names with the current phylogenetic classification

# Reformer ScientificNameCor
Si "NA NA" -> NA_character_
```{r}
Data[, ScientificNameCor := paste(GenusCor, SpeciesCor)]

Data[, ScientificNameCor := ifelse(ScientificNameCor == "NA NA", NA_character_, ScientificNameCor)]

```

